rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,1222 @@
+users to inject their own special logic within the normal tuf installation flow.
+
+#### Remote-Script Install Controller
+
+- Affects installation indirectly by running a user-provided script on a pre-determined host
+(not the host in need of upgrade).
+
+- Intended as a simple means of hooking into systems such as k8s, where the teleport version is
+controlled via a remote API, though that does not preclude us making official remote install
+controllers for specific APIs down the road.
+
+- Details of functionality are TBD, but the basic idea will be that we will mirror the functionality of
+`local-script` wherever possible, and add an additional server selector that is used to
+determine where the installer should be run.
+
+- Q: Should the list of target servers be provided to the script?  Is that even useful?  It seems
+more likely that scripts will be written per externally managed set, though that could be a
+failure of imagination on my part.
+
+
+### TUF CI and Repository
+
+In order to enable the TUF version controller, we will need to maintain CI that generates and signs
+TUF metadata, and maintain a TUF repository.  Details of how the TUF repository will be hosted are
+still TBD, but TUF repositories are basically static files, so distribution should be fairly straightforward.
+We may be able to simply distribute it via a git repo.
+
+We will leverage deterministic builds and TUF's multisignature support to harden ourselves against CI
+compromise. Our standard build pipeline will generate and sign one set of package hashes, and another set
+will be generated and signed by a separate isolated env.
+
+TUF repositories prove liveness via periodic resigning with a "hot" key (not the keys used for package signing).
+This hot key should be isolated from the package signing keys, so we're likely looking at two new isolated
+envs that need to be added in addition to the modifications to our existing CI.
+
+*note*: some initial work was done to get deterministic builds working on linux packages. We know its possible
+(and might even still be working), but don't currently have test coverage for build determinism. This will be
+an important part of the prerequisite work to get the TUF system online. We don't need to add TUF support for
+all build targets at once, so we may specifically target reliable signing of amd64/linux packages first.
+
+
+### Rollbacks
+
+Rollbacks will come in two flavors:
+
+1. Remote rollback: Version directive is changed to target an older version. Older version is installed via
+normal install controller. Requires the new teleport installation to work at least well
+enough to perform any functions required by the install controller.
+
+2. Local rollback: The previous teleport installation remains cached during the upgrade, and some local process
+monitors the health of the new version. If the new version remains unhealthy for too long, it is forcibly
+terminated and the previous installation is replaced.
```

Comment:
Is this planned to expand or replace the current readiness check for the new process during HUP?